src/llama-kv-cache.cpp

```diff
@@ -438,23 +567,63 @@ bool llama_kv_cache_unified::find_slot(const llama_ubatch & ubatch) {
         }
         LLAMA_LOG_WARN("\n%s\n", ss.c_str());
     }
+
+    LLAMA_LOG_WARN("kv_cells: n_swa = %4d, min[0] = %5d, max[0] = %5d\n", n_swa, cells.seq_pos_min(0), cells.seq_pos_max(0));
+    LLAMA_LOG_WARN("kv_cells: n_swa = %4d, min[1] = %5d, max[1] = %5d\n", n_swa, cells.seq_pos_min(1), cells.seq_pos_max(1));
+    LLAMA_LOG_WARN("kv_cells: n_swa = %4d, min[2] = %5d, max[2] = %5d\n", n_swa, cells.seq_pos_min(2), cells.seq_pos_max(2));
 #endif
 
     uint32_t n_tested = 0;
 
     while (true) {
-        if (head + n_tokens > cells.size()) {
-            n_tested += cells.size() - head;
-            head = 0;
+        if (head_cur + n_tokens > cells.size()) {
+            n_tested += cells.size() - head_cur;
+            head_cur = 0;
             continue;
         }
 
+        // keep track of what the minimum sequence positions would be if we accept the ubatch
+        llama_seq_id seq_pos_min[LLAMA_MAX_PARALLEL_SEQUENCES];
+        for (int s = 0; s < LLAMA_MAX_PARALLEL_SEQUENCES; ++s) {
+            seq_pos_min[s] = cells.seq_pos_min(s);
+        }
+
         bool found = true;
         for (uint32_t i = 0; i < n_tokens; i++) {
-            // TODO: improve to accept cells that are masked by the SWA
-            if (!cells.is_empty(head + i)) {
+            const llama_pos    pos    = ubatch.pos[i];
+            const llama_seq_id seq_id = ubatch.seq_id[i][0];
```

Comment:
In local testing trying to use `split_equal` with the unified portion of the hybrid cache, this line is causing problems. Specifically, I think it conflicts with the logic in `llama_sbatch::add_seq_to_ubatch` ([here](https://github.com/ggml-org/llama.cpp/blob/gg/kv-cache-simplify-part3/src/llama-batch.cpp#L90)) where the `ubatch.seq_id` is only populated with a non-null value at position `ubatch.n_seqs`. I'll keep digging to see if there's a simple solution, but wanted to flag this in case it's an easy fix on your end.

Reply:
This is probably not the _right_ solution, but this "fixed" the issue on my Granite 4 branch:
```suggestion
            const llama_seq_id seq_id = ubatch.seq_id[i] == nullptr ? ubatch.seq_id[0][0] : ubatch.seq_id[i][0];

```

Reply:
> I'm not quite following the relationship that should be used to extract `seq_id` for the given `pos` element.

Usually, this is relatively simple since the number of tokens per sequence is known in `ubatch.n_seq_tokens`. In fact, here it could probably be possible to use

```suggestion
            const llama_seq_id seq_id = ubatch.seq_id[i / ubatch.n_seq_tokens][0];
```

although there is another approach without divisions, but with nested loops and which would change the indexing for `ubatch.pos[i]` to `ubatch.pos[s * ubatch.n_seq_tokens + j]` where `s` is in `[0, ubatch.n_seqs)` and `j` in `[0, ubatch.n_seq_tokens)`.